quantumdb-postgresql/src/main/java/io/quantumdb/core/planner/PostgresqlMigrator.java

```diff
@@ -124,15 +122,61 @@ public void drop(State state, Version version) throws MigrationException {
 		RefLog refLog = state.getRefLog();
 		Catalog catalog = state.getCatalog();
 
-		List<TableRef> tablesToDrop = refLog.getTableRefs().stream()
-				.filter(tableRef -> tableRef.getVersions().contains(version))
-				.filter(tableRef -> tableRef.getVersions().stream()
-						.filter(otherVersion -> otherVersion.getOperation().getType() == Type.DDL)
-						.count() == 1)
-				.collect(Collectors.toList());
+		Set<Version> activeVersions = refLog.getVersions();
+		checkArgument(activeVersions.contains(version), "Version to drop is not part of the active versions!");
+
+		// Check if version is at the end of the active versions
+		Version first = version;
+		Version last = version;
+		Version intermediateV = version.getChild();
+		while (intermediateV != null) {
+			if (activeVersions.contains(intermediateV)) {
+				last = intermediateV;
+			}
+			intermediateV = intermediateV.getChild();
+		}
+		// Or first
+		intermediateV = version.getParent();
+		while (intermediateV != null) {
+			if (activeVersions.contains(intermediateV)) {
+				first = intermediateV;
+			}
+			intermediateV = intermediateV.getParent();
+		}
+
+		checkArgument(!(first.equals(version) && last.equals(version)), "Cannot drop the only active version!");
```

Comment:
I could be wrong, but I think it's possible to have multiple active versions that are not directly a child/parent of one another, but instead have an intermediate version that is not active. Examples of this would be a base version, and then having applied a migration using a changeset with >1 operations. Only the base version and the last operation's version would be marked as active.

Reply:
Hmm, this one is indeed difficult to solve as QuantumDB does not show the user any intermediate versions (only the last one from a changeset). I had a problem where I added 7 foreign key constraint in 1 changeset and when I wanted to drop the version that added it, there were 2 "ghost" tables that were not removed. See the changesets below.

I think that users should keep in mind that 1 changeset should be considered 1 atomic change. If they want to reverse the change, it is either every change in that changeset or none.

I have changed the drop functionality between now and when I submitted this pull request because I noticed an error. 

I will think about what the exact drop functionality should be. If you take a look at the documentation at [https://quantumdb.io/docs/master/#changesets](https://quantumdb.io/docs/master/#changesets), whenever you drop the "oldest" version, in this case 8fb3b4e71a, all the intermediate versions until the next active version (cf3323a79b) should also be dropped (tables will only be dropped if they are not part of the next active version).

If the newest active version gets dropped, also the versions in the changeset to get to the newest version should be dropped, in the example this would be both 72ae28b49a and cc92b7b121. However, if you drop an intermediate active version, what would the functionality be? You use the intermediate versions to get to the next versions, so if you want to drop that one, how can you get to the newest versions? I think the functionality, in this case, should be to only drop the version from the active list and not drop any tables until they are part of either the oldest or newest version to be dropped.

The below changesets can be tested on a database made with [HammerDB](https://www.hammerdb.com/download.html)
```
  <changeset id="add_foreignkeys" author="Nick Richter">
    <description>Add the foreignkey constrains</description>
    <operations>
		<addForeignKey tableName="district" columnNames="d_w_id" referencesTableName="warehouse" referencesColumnNames="w_id" name="district_warehouse_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="customer" columnNames="c_w_id,c_d_id" referencesTableName="district" referencesColumnNames="d_w_id,d_id" name="customer_district_id_fk" onDelete="CASCADE" />
    </operations>
  </changeset>
  
  <changeset id="add_other_foreignkeys" author="Nick Richter">
    <description>Add all other foreignkeys</description>
    <operations>
		<addForeignKey tableName="history" columnNames="h_w_id,h_d_id" referencesTableName="district" referencesColumnNames="d_w_id,d_id" name="history_district_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="orders" columnNames="o_w_id,o_d_id,o_c_id" referencesTableName="customer" referencesColumnNames="c_w_id,c_d_id,c_id" name="orders_customers_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="new_order" columnNames="no_w_id,no_d_id,no_o_id" referencesTableName="orders" referencesColumnNames="o_w_id,o_d_id,o_id" name="neworder_orders_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="stock" columnNames="s_w_id" referencesTableName="warehouse" referencesColumnNames="w_id" name="stock_warehouse_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="stock" columnNames="s_i_id" referencesTableName="item" referencesColumnNames="i_id" name="stock_item_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="order_line" columnNames="ol_w_id,ol_d_id,ol_o_id" referencesTableName="orders" referencesColumnNames="o_w_id,o_d_id,o_id" name="orderline_orders_id_fk" onDelete="CASCADE" />
		<addForeignKey tableName="order_line" columnNames="ol_supply_w_id,ol_i_id" referencesTableName="stock" referencesColumnNames="s_w_id,s_i_id" name="orderline_stock_id_fk" onDelete="CASCADE" />
    </operations>

  </changeset>
  <changeset id="add_column" author="Nick Richter">
    <description>Add a new tax column to order_line</description>
    <operations>
		<addColumn tableName="order_line">
        <column name="ol_tax" type="numeric" defaultExpression="0" nullable="false" />
      </addColumn>
    </operations>
  </changeset>
```

Reply:
> Just so I can reproduce this:
> 
>     Use HammerDB as the baseline
>     Run quantumdb init to index the database.
>     Add this changelog.xml file.
>     Run quantumdb fork <version for "add_column">
>     Run quantumdb drop <root version>
>     End up with 2 ghost tables that should've been removed.
> 
> Is this correct? And do you happen to know which tables these 2 ghost tables represented?

Yes that is correct, the only thing I did differently was to fork each changeset instead of only the last one.
I will test it again myself to see if it also happens when I only specify the last one.

I have only tested QuantumDB by forking 1 changeset at a time, that is probably the reason why I think about it differently.

Reply:
You cannot drop the initial version in QuantumDB version 0.4.1-Snapshot as you get a NullPointerException when calling getType() on the root operation (which is null):
```
	@Override
	public void drop(State state, Version version) throws MigrationException {
		// Check that the version's operation is of type DDL, or has no operation (root version).
		checkArgument(version.getOperation() == null || version.getOperation().getType() == Type.DDL);

		RefLog refLog = state.getRefLog();
		Catalog catalog = state.getCatalog();

		List<TableRef> tablesToDrop = refLog.getTableRefs().stream()
				.filter(tableRef -> tableRef.getVersions().contains(version))
				.filter(tableRef -> tableRef.getVersions().stream()
						.filter(otherVersion -> **otherVersion.getOperation().getType()** == Type.DDL)
						.count() == 1)
				.collect(Collectors.toList());
```

Reply:
Using the "old" drop logic and what you above described it only deleted the 'district' table, the rest of the tables were not dropped when dropping the initial state.

Reply:
I'll double check myself as well, as you should be able to drop the initial version after the first fork. Will let you know what I find.

Reply:
This one is with my new drop logic which is correct:
![2021-05-12 17_11_09-pgAdmin 4 — Mozilla Firefox](https://user-images.githubusercontent.com/22129925/117999438-22080b80-b345-11eb-9130-dab8f80072d0.png)

Reply:
Managed to set up a database with the HammerDB schema yesterday, and ran into the `numeric` issue when trying to init quantumdb. I'll use some of the fixes from this PR to check what's going wrong. I'll keep you updated.